### DISCUSSION TITLE
Filtering INotificationHandler when retrying

### DISCUSSION THREAD

**Discussion by anthony-keller** [LuckyPennySoftware/MediatR#827] [Q&A]
**Feature Request**

I have an pub/sub system and the subscriber retrieves an event from a queue and publishes an INotification. If all handlers execute successfully, the item is removed from the queue. 

In cases where there are multiple handlers, if any of the handlers throw an exception, the item is not removed from the queue. A short time later, the same item is retrieved and the notification is republished as a retry. This executes all handlers again, however, I would like a way to skip any that were executed successfully in previous attempts. 

The INotification has an event id that would be combined with the type name of the handler to create a unique identifier for the execution of the handler per event. 

It would be great to be able to execute some code before each handler is executed to check some sort of data source for whether the handler has been executed for the event id. 

I wanted to get your thoughts on this idea first, and if you think it has some merit, I will take a closer look at how it might be implemented. I feel like it's similar to a pipeline behaviour.

**Comment by anthony-keller** [comment]
I can certainly inject into each handler and exit early if the event has already been successfully handled, but I'm interested in implementing something so the developer doesn't have to add the same code in each handler.

**Comment by jbogard** [comment]
The right way to handle this I think is to make each of your notification handlers an actual subscriber to the event itself. You're blurring the line between queuing and in-process handling, and MediatR is NOT the appropriate solution for that. I would create distinct subscribers for that queue message and have each subscriber manage its own retries/idempotency. NServiceBus/MassTransit etc are the solution I would use.
